cpp/src/join/mixed_join_common_utils.cuh

```diff
@@ -176,6 +177,97 @@ using hash_set_type =
 // The hash_set_ref_type used by mixed_semi_join kerenels for probing.
 using hash_set_ref_type = hash_set_type::ref_type<cuco::contains_tag>;
 
+/**
+ * @brief Common utility for probing a hash table bucket and checking slot equality
+ *
+ * This encapsulates the common logic of reading bucket slots and checking for
+ * empty slots and key equality, used by both count and retrieve operations.
+ */
+template <bool has_nulls>
+struct hash_probe_result {
+  bool first_slot_is_empty;
+  bool second_slot_is_empty;
+  bool first_slot_equals;
+  bool second_slot_equals;
+
+  __device__ __forceinline__ hash_probe_result(
+    pair_expression_equality<has_nulls> const& key_equal,
+    cudf::device_span<cuco::pair<hash_value_type, cudf::size_type>> hash_table_storage,
+    cuco::pair<hash_value_type, cudf::size_type> const& probe_key,
+    std::size_t probe_idx)
+  {
+    auto const* data = hash_table_storage.data();
+    auto const bucket_slots =
+      *reinterpret_cast<cuda::std::array<cuco::pair<hash_value_type, cudf::size_type>, 2> const*>(
+        data + probe_idx);
```

Comment:
This will likely not emit a vector load, since the `pair` and thus the `array` inherit the alignment of the underlying `hash_value_type`. I you can guarantee alignment, you can do something like

```suggestion
    auto const bucket_slots =
      *reinterpret_cast<cuda::std::array<cuco::pair<hash_value_type, cudf::size_type>, 2> const*>(
        __builtin_assume_aligned(data + probe_idx, sizeof(hash_value_type) * 2));
```

Reply:
This is nicer. And indeed, the alignment assumption is really the thing that matters. If we do that, we don't need the reinterpret cast at all:

https://godbolt.org/z/M9jMfnP64

If we do this, we should `CUDF_EXPECTS` the alignment of this argument _before_ launching the kernel. And we need to make sure that every invocation of this function does pass off an appropriately aligned device_span.

Reply:
We perform the casting primarily because the vector length isn’t fixed at 2 in cuco, so the manual unrolling shown in your Godbolt example would have to be implemented as a loop. The goal here is simply to help the compiler emit the widest possible load instruction. In this specific case, your Godbolt example provides a very clean solution since the vector length is known to be 2. Casting isn’t needed here. Let me update the code accordingly and we should adopt this approach in cuco as well.